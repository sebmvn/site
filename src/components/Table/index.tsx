import DataTable, { TableColumn } from 'react-data-table-component'

export type DataRow = Record<string, string>
export type TableProps = {
  columns: string[]
  data: DataRow[]
}

const Table = ({ columns, data }: TableProps) => {
  const keys = Object.keys(data[0])
  const tableColumns: TableColumn<DataRow>[] = keys.map((key, index) => {
    return {
      name: columns[index],
      selector: (row) => row[key]
    }
  })

  return <DataTable columns={tableColumns} data={data} />
}

export default Table
