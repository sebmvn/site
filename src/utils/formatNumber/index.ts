export const formatPriceNotation = (value: string) => {
  value = value?.replace('.', '').replace(',', '').replace(/\D/g, '')

  if (isNaN(parseFloat(value))) {
    return ''
  }

  const options = { minimumFractionDigits: 2 }
  const result = new Intl.NumberFormat('pt-BR', options).format(
    parseFloat(value) / 100
  )

  return result
}

export const formatIntegerNotation = (value: string) => {
  return value?.replace('.', '').replace(',', '').replace(/\D/g, '')
}

export const toDecimal = (value: string) => {
  const formattedValue = formatIntegerNotation(value)
  return parseFloat(formattedValue!) / 100
}

export const toInteger = (value: string) => {
  const formattedValue = formatIntegerNotation(value)
  return parseInt(formattedValue!)
}
