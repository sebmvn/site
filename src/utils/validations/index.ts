import { CalendarType } from 'hooks/use-calc'
import Joi from 'joi'

const fieldsValidations = {
  initialValue: Joi.string().required().messages({
    'any.required': 'Campo obrigatório',
    'string.empty': 'Campo obrigatório',
    'string.min': 'tamanho minimo de 1 digito',
    'string.base': 'o tipo do dado precisa ser textual'
  }),
  monthlyValue: Joi.string().required().messages({
    'any.required': 'Campo obrigatório',
    'string.empty': 'Campo obrigatório',
    'string.min': 'tamanho minimo de 1 digito',
    'string.base': 'o tipo do dado precisa ser textual'
  }),

  interestRate: Joi.string().required().messages({
    'any.required': 'Campo obrigatório',
    'string.empty': 'Campo obrigatório',
    'string.min': 'tamanho minimo de 1 digito',
    'string.base': 'o tipo do dado precisa ser textual'
  }),
  period: Joi.string().required().messages({
    'any.required': 'Campo obrigatório',
    'string.empty': 'Campo obrigatório',
    'string.min': 'tamanho minimo de 1 digito',
    'string.base': 'o tipo do dado precisa ser textual'
  }),
  interestRateCalendar: Joi.string().valid('monthly', 'yearly').required(),
  periodCalendar: Joi.string().valid('monthly', 'yearly').required()
}

export type FieldErrors = {
  [key: string]: string
}

const getFieldErrors = (objError: Joi.ValidationResult) => {
  const errors: FieldErrors = {}

  if (objError.error) {
    objError.error.details.forEach((err) => {
      errors[err.path.join('.')] = err.message
    })
  }

  return errors
}

type compoundInterestCalculatorFields = {
  initialValue: string
  monthlyValue: string
  interestRate: string
  period: string
  interestRateCalendar: CalendarType
  periodCalendar: CalendarType
}
export const compoundInterestCalculatorValidate = (
  values: compoundInterestCalculatorFields
) => {
  const schema = Joi.object(fieldsValidations)

  return getFieldErrors(schema.validate(values, { abortEarly: false }))
}
