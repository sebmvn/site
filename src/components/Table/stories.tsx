import { Story, Meta } from '@storybook/react/types-6-0'
import Table, { TableProps } from '.'

export default {
  title: 'Table',
  component: Table,
  args: {
    columns: [
      'MÃªs',
      'Juros',
      'Total Investido',
      'Total Juros',
      'Total Acumulado'
    ],
    data: [
      {
        month: '0',
        interestRateMonthly: '--',
        totalValue: 'R$ 5.000,0',
        totalInInterest: 'R$ 0,00',
        totalAmountInvested: 'R$ 5.000,00'
      },
      {
        month: '1',
        interestRateMonthly: 'R$ 50,00',
        totalValue: 'R$ 7.000,0',
        totalInInterest: 'R$ 50,00',
        totalAmountInvested: 'R$ 7.050,00'
      },
      {
        month: '2',
        interestRateMonthly: 'R$ 70,50',
        totalValue: 'R$ 9.000,0',
        totalInInterest: 'R$ 120,50',
        totalAmountInvested: 'R$ 9.120,50'
      }
    ]
  }
} as Meta

export const Default: Story<TableProps> = (args) => <Table {...args} />
