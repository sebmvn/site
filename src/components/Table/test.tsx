import { screen } from '@testing-library/react'
import { renderWithTheme } from 'utils/test-utils'

import Table from '.'

describe('<Table />', () => {
  const props = {
    columns: [
      'Mês',
      'Juros',
      'Total Investido',
      'Total Juros',
      'Total Acumulado'
    ],
    data: [
      {
        month: '0',
        interestRateMonthly: '--',
        totalValue: '5.000,0',
        totalInInterest: '0,00',
        totalAmountInvested: '5.000,00'
      },
      {
        month: '1',
        interestRateMonthly: '50,00',
        totalValue: '7.000,0',
        totalInInterest: '50,00',
        totalAmountInvested: '7.050,00'
      },
      {
        month: '2',
        interestRateMonthly: '70,50',
        totalValue: '9.000,0',
        totalInInterest: '120,50',
        totalAmountInvested: '9.120,50'
      }
    ]
  }

  it('should render the table', () => {
    const { container } = renderWithTheme(<Table {...props} />)

    expect(
      screen.getByRole('columnheader', { name: /mês/i })
    ).toBeInTheDocument()
    expect(
      screen.getByRole('row', {
        name: /Mês Juros Total Investido Total Juros Total Acumulado/i
      })
    ).toBeInTheDocument()

    expect(container.firstChild).toMatchSnapshot()
  })
})
