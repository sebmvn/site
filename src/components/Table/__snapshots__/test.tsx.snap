// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Table /> should render the table 1`] = `
.c2 {
  position: relative;
  box-sizing: border-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  width: 100%;
  height: 100%;
  max-width: 100%;
  color: rgba(0,0,0,0.87);
  background-color: #FFFFFF;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: 100%;
  color: rgba(0,0,0,0.87);
  font-size: 12px;
  font-weight: 500;
}

.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: stretch;
  -webkit-box-align: stretch;
  -ms-flex-align: stretch;
  align-items: stretch;
  width: 100%;
  background-color: #FFFFFF;
  min-height: 52px;
  border-bottom-width: 1px;
  border-bottom-color: rgba(0,0,0,.12);
  border-bottom-style: solid;
}

.c5 {
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  box-sizing: border-box;
  line-height: normal;
  padding-left: 16px;
  padding-right: 16px;
}

.c11 {
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  box-sizing: border-box;
  line-height: normal;
  padding-left: 16px;
  padding-right: 16px;
  word-break: break-word;
}

.c6 {
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  -webkit-flex-basis: 0;
  -ms-flex-preferred-size: 0;
  flex-basis: 0;
  max-width: 100%;
  min-width: 100px;
}

.c12 div:first-child {
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
}

.c10 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: stretch;
  -webkit-box-align: stretch;
  -ms-flex-align: stretch;
  align-items: stretch;
  -webkit-align-content: stretch;
  -ms-flex-line-pack: stretch;
  align-content: stretch;
  width: 100%;
  box-sizing: border-box;
  font-size: 13px;
  font-weight: 400;
  color: rgba(0,0,0,0.87);
  background-color: #FFFFFF;
  min-height: 48px;
}

.c10:not(:last-of-type) {
  border-bottom-style: solid;
  border-bottom-width: 1px;
  border-bottom-color: rgba(0,0,0,.12);
}

.c7 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: inherit;
  -webkit-justify-content: inherit;
  -ms-flex-pack: inherit;
  justify-content: inherit;
  height: 100%;
  width: 100%;
  outline: none;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  overflow: hidden;
}

.c8 {
  overflow: hidden;
  white-space: nowrap;
  text-overflow: ellipsis;
}

.c9 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c0 {
  position: relative;
  width: 100%;
  border-radius: inherit;
  overflow-x: auto;
  overflow-y: hidden;
  min-height: 0;
}

.c1 {
  position: relative;
  width: 100%;
  display: table;
}

<div
  class="c0"
>
  <div
    class="c1"
  >
    <div
      class="c2 rdt_Table"
      role="table"
    >
      <div
        class="c3 rdt_TableHead"
        role="rowgroup"
      >
        <div
          class="c4 rdt_TableHeadRow"
          role="row"
        >
          <div
            class="c5 c6 rdt_TableCol"
            data-column-id="1"
          >
            <div
              class="c7 rdt_TableCol_Sortable"
              data-column-id="1"
              data-sort-id="1"
              disabled=""
              role="columnheader"
              tabindex="0"
            >
              <div
                class="c8"
                data-column-id="1"
              >
                MÃªs
              </div>
            </div>
          </div>
          <div
            class="c5 c6 rdt_TableCol"
            data-column-id="2"
          >
            <div
              class="c7 rdt_TableCol_Sortable"
              data-column-id="2"
              data-sort-id="2"
              disabled=""
              role="columnheader"
              tabindex="0"
            >
              <div
                class="c8"
                data-column-id="2"
              >
                Juros
              </div>
            </div>
          </div>
          <div
            class="c5 c6 rdt_TableCol"
            data-column-id="3"
          >
            <div
              class="c7 rdt_TableCol_Sortable"
              data-column-id="3"
              data-sort-id="3"
              disabled=""
              role="columnheader"
              tabindex="0"
            >
              <div
                class="c8"
                data-column-id="3"
              >
                Total Investido
              </div>
            </div>
          </div>
          <div
            class="c5 c6 rdt_TableCol"
            data-column-id="4"
          >
            <div
              class="c7 rdt_TableCol_Sortable"
              data-column-id="4"
              data-sort-id="4"
              disabled=""
              role="columnheader"
              tabindex="0"
            >
              <div
                class="c8"
                data-column-id="4"
              >
                Total Juros
              </div>
            </div>
          </div>
          <div
            class="c5 c6 rdt_TableCol"
            data-column-id="5"
          >
            <div
              class="c7 rdt_TableCol_Sortable"
              data-column-id="5"
              data-sort-id="5"
              disabled=""
              role="columnheader"
              tabindex="0"
            >
              <div
                class="c8"
                data-column-id="5"
              >
                Total Acumulado
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="c9 rdt_TableBody"
        role="rowgroup"
      >
        <div
          class="c10 rdt_TableRow"
          id="row-0"
          role="row"
        >
          <div
            class="c11 c6 c12 rdt_TableCell"
            data-column-id="1"
            data-tag="allowRowEvents"
            id="cell-1-undefined"
            role="cell"
          >
            <div
              data-tag="allowRowEvents"
            >
              0
            </div>
          </div>
          <div
            class="c11 c6 c12 rdt_TableCell"
            data-column-id="2"
            data-tag="allowRowEvents"
            id="cell-2-undefined"
            role="cell"
          >
            <div
              data-tag="allowRowEvents"
            >
              --
            </div>
          </div>
          <div
            class="c11 c6 c12 rdt_TableCell"
            data-column-id="3"
            data-tag="allowRowEvents"
            id="cell-3-undefined"
            role="cell"
          >
            <div
              data-tag="allowRowEvents"
            >
              5.000,0
            </div>
          </div>
          <div
            class="c11 c6 c12 rdt_TableCell"
            data-column-id="4"
            data-tag="allowRowEvents"
            id="cell-4-undefined"
            role="cell"
          >
            <div
              data-tag="allowRowEvents"
            >
              0,00
            </div>
          </div>
          <div
            class="c11 c6 c12 rdt_TableCell"
            data-column-id="5"
            data-tag="allowRowEvents"
            id="cell-5-undefined"
            role="cell"
          >
            <div
              data-tag="allowRowEvents"
            >
              5.000,00
            </div>
          </div>
        </div>
        <div
          class="c10 rdt_TableRow"
          id="row-1"
          role="row"
        >
          <div
            class="c11 c6 c12 rdt_TableCell"
            data-column-id="1"
            data-tag="allowRowEvents"
            id="cell-1-undefined"
            role="cell"
          >
            <div
              data-tag="allowRowEvents"
            >
              1
            </div>
          </div>
          <div
            class="c11 c6 c12 rdt_TableCell"
            data-column-id="2"
            data-tag="allowRowEvents"
            id="cell-2-undefined"
            role="cell"
          >
            <div
              data-tag="allowRowEvents"
            >
              50,00
            </div>
          </div>
          <div
            class="c11 c6 c12 rdt_TableCell"
            data-column-id="3"
            data-tag="allowRowEvents"
            id="cell-3-undefined"
            role="cell"
          >
            <div
              data-tag="allowRowEvents"
            >
              7.000,0
            </div>
          </div>
          <div
            class="c11 c6 c12 rdt_TableCell"
            data-column-id="4"
            data-tag="allowRowEvents"
            id="cell-4-undefined"
            role="cell"
          >
            <div
              data-tag="allowRowEvents"
            >
              50,00
            </div>
          </div>
          <div
            class="c11 c6 c12 rdt_TableCell"
            data-column-id="5"
            data-tag="allowRowEvents"
            id="cell-5-undefined"
            role="cell"
          >
            <div
              data-tag="allowRowEvents"
            >
              7.050,00
            </div>
          </div>
        </div>
        <div
          class="c10 rdt_TableRow"
          id="row-2"
          role="row"
        >
          <div
            class="c11 c6 c12 rdt_TableCell"
            data-column-id="1"
            data-tag="allowRowEvents"
            id="cell-1-undefined"
            role="cell"
          >
            <div
              data-tag="allowRowEvents"
            >
              2
            </div>
          </div>
          <div
            class="c11 c6 c12 rdt_TableCell"
            data-column-id="2"
            data-tag="allowRowEvents"
            id="cell-2-undefined"
            role="cell"
          >
            <div
              data-tag="allowRowEvents"
            >
              70,50
            </div>
          </div>
          <div
            class="c11 c6 c12 rdt_TableCell"
            data-column-id="3"
            data-tag="allowRowEvents"
            id="cell-3-undefined"
            role="cell"
          >
            <div
              data-tag="allowRowEvents"
            >
              9.000,0
            </div>
          </div>
          <div
            class="c11 c6 c12 rdt_TableCell"
            data-column-id="4"
            data-tag="allowRowEvents"
            id="cell-4-undefined"
            role="cell"
          >
            <div
              data-tag="allowRowEvents"
            >
              120,50
            </div>
          </div>
          <div
            class="c11 c6 c12 rdt_TableCell"
            data-column-id="5"
            data-tag="allowRowEvents"
            id="cell-5-undefined"
            role="cell"
          >
            <div
              data-tag="allowRowEvents"
            >
              9.120,50
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
